### ISSUE TITLE
How to save dialog progress to LocalStorage?

### ISSUE THREAD

**Issue by Iilinkasa** [dmitrizzle/chat-bubble#91]
Hi, sorry for my english
How i can save dialog progress to LocalStorage?

**Comment by dmitrizzle** [comment]
Hey @Iilinkasa, sorry I haven't documented the API well enough here. You can set the number of interactions to save in `localStorage` automatically by setting `options. recallInteractions ` to a number value: https://github.com/dmitrizzle/chat-bubble/blob/master/component/Bubbles.js#L9

For example,
```javascript
var chatWindow = new Bubbles(
  document.getElementById("chat"),
  "chatWindow",
  { recallInteractions: 20 }
)
```

**Comment by Iilinkasa** [comment]
It saves great, but how can I download progress now? After all, after the reboot, everything exactly everything happens from a new stage.

**Comment by Iilinkasa** [comment]
the story just doesn't start from the right moment
the history of the dialogue is displayed but there is no continuation of it
Example
![image](https://user-images.githubusercontent.com/92829677/138547564-3b87b79f-8243-47e0-bce1-f644807b7873.png)


i set 
Bubble.js
 recallInteractions = options.recallInteractions || 20 // number of interactions to be remembered and brought back upon restart

is it possible to make the story go on?

**Comment by dmitrizzle** [comment]
Hi, you should be able to, yes. Quoting my response from Twitter:

Hey have you tried combining that with a custom starting point? There’s an example here https://github.com/dmitrizzle/chat-bubble/blob/master/examples/2-custom-starting-point.html

Essentially, you’re initializing the chat restart like this:

```
// CUSTOM STARTPOINT for conversation block defined as a
// second parameter in .talk() function:
chatWindow.talk(convo, "banana")
```

**Comment by Iilinkasa** [comment]
```
<!-- import the JavaScript file -->
<script src="../component/Bubbles.js"></script>
<script>
// initialize by constructing a named function...
var chatWindow = new Bubbles(
  document.getElementById("chat"), // ...passing HTML container element...
  "chatWindow" // ...and name of the function as a parameter
    { recallInteractions: 20 }
)

// conversation object defined separately, but just the same as in the
// "Basic chat-bubble Example" (1-basics.html)
var convo = {
  // "ice" (as in "breaking the ice") is a required conversation object
  // that maps the first thing the bot will say to the user
  ice: {
    // "says" defines an array of sequential bubbles
    // that the bot will produce
    says: [
		"<span class=\"avataryes, I initiate the restart, but still the chat starts again and not from the moment of the restart.
\"></span>Тут есть кто-то?.",
		"Ау",

	],

    // "reply" is an array of possible options the user can pick from
    // as a reply
    reply: [
      {
        question: "Кто это?", // label for the reply option
        answer: "who_is" // key for the next conversation object
      }
    ]
  }, // end required "ice" conversation object

  // another conversation object that can be queued from within
  // any other conversation object, including itself
  who_is: {
    says: ["Алиса но какая разница?"],
    reply: [
      {
        question: "Привет",
        answer: "ice"

      }
    ]
  } // end conversation object
}

// CUSTOM STARTPOINT for conversation block defined as a
// second parameter in .talk() function:
chatWindow.talk(convo,"ice" )

```


I tried to separately store the responses in a function and create them in localStorage, but this is a very long job for each response

```
<script src="../component/Bubbles.js"></script>
<script>
	var convo = {
				q1: {
					says: [
						"Меня слышно?!",
						"Помогите кто нибудь",
						"Где же вы...",
					],
					reply: [
						{
							question: 'ответ 1',
							answer: 'func1'
						},
						{
							question: 'ответ 2',
							answer: 'func2'
						}
					]
				},
				q2: {
					says: [
						"Вопрос 2",
					],
					reply: [
						{
							question: 'ответ 1',
							answer: 'func3'
						},
						{
							question: 'ответ 2',
							answer: 'func4'
						}
					]
				},
				q3: {
					says: [
						"Вопрос 3",
					],
					reply: [
						{
							question: 'ответ 1',
							answer: 'func5'
						},
						{
							question: 'ответ 2',
							answer: 'func6'
						}
					]
				}
			},

			func1 = function() {
				localStorage.chatProgress = 'q2'
				chatWindow.talk(convo, "q2")
			}

	func3 = function() {
		localStorage.chatProgress = 'q3'
		chatWindow.talk(convo, "q3")
	}

	var chatWindow = new Bubbles(
			document.getElementById("chat"), // ...passing HTML container element...
			"chatWindow" // ...and name of the function as a parameter
	)
	const chatProgress = localStorage.getItem('chatProgress') || 'q1'
	chatWindow.talk(convo, chatProgress)
</script>
```

**Comment by Iilinkasa** [comment]
What you think?

**Comment by dmitrizzle** [comment]
In your implementation, you made a few syntax errors. Please check your browser console for errors first to figure out what's going on. I fixed your code and it seems to work:

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Basic chat-bubble Example</title>

    <!-- for mobile screens -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />

    <!-- stylesheets are conveniently separated into components -->
    <link rel="stylesheet" media="all" href="../component/styles/setup.css" />
    <link rel="stylesheet" media="all" href="../component/styles/says.css" />
    <link rel="stylesheet" media="all" href="../component/styles/reply.css" />
    <link rel="stylesheet" media="all" href="../component/styles/typing.css" />
    <link rel="stylesheet" media="all" href="../component/styles/input.css" />
    <style>
      body {
        background: #dcdde0;
      }
      .bubble-container {
        height: 100vh;
      }
    </style>
  </head>
  <body>
    <!-- container element for chat window -->
    <div id="chat"></div>

    <!-- import the JavaScript file -->
    <script src="../component/Bubbles.js"></script>
    <script>
      // initialize by constructing a named function...
      var chatWindow = new Bubbles(
        document.getElementById("chat"), // ...passing HTML container element...
        "chatWindow", // ...and name of the function as a parameter
        { recallInteractions: 20 }
      );

      // conversation object defined separately, but just the same as in the
      // "Basic chat-bubble Example" (1-basics.html)
      var convo = {
        // "ice" (as in "breaking the ice") is a required conversation object
        // that maps the first thing the bot will say to the user
        ice: {
          // "says" defines an array of sequential bubbles
          // that the bot will produce
          says: [
            '<span class="avataryes, I initiate the restart, but still the chat starts again and not from the moment of the restart."></span>Тут есть кто-то?.',
            "Ау",
          ],

          // "reply" is an array of possible options the user can pick from
          // as a reply
          reply: [
            {
              question: "Кто это?", // label for the reply option
              answer: "who_is", // key for the next conversation object
            },
          ],
        }, // end required "ice" conversation object

        // another conversation object that can be queued from within
        // any other conversation object, including itself
        who_is: {
          says: ["Алиса но какая разница?"],
          reply: [
            {
              question: "Привет",
              answer: "ice",
            },
          ],
        }, // end conversation object
      };

      // CUSTOM STARTPOINT for conversation block defined as a
      // second parameter in .talk() function:
      chatWindow.talk(convo, "ice");
    </script>
  </body>
</html>

```

**Comment by Iilinkasa** [comment]
Русский
Английский
я скопировал код.
запускаю: 
Да, я вижу историю чата, но при перезагрузке окна, диалог начинается сначала.
и не продолжается. К примеру
я выбрал 2 кнопки ответа с момента запуска, а после перезагрузки окна
диалог начинается с первого вопроса.

Возможно вы сможете указать пример, где это работает корректно, что-бы после перезапуска, диалог продолжался с момента перезагрузки.
 



ya skopiroval kod.
zapuskayu: 
Da, ya vizhu istoriyu chata, no pri perezagruzke okna, dialog nachinayetsya snachala.
i ne prodolzhayetsya. K primeru
ya vybral 2 knopki otveta s momenta zapuska, a posle perezagruzki okna
dialog nachinayetsya s pervogo voprosa.

Vozmozhno vy smozhete ukazat' primer, gde eto rabotayet korrektno, chto-by posle perezapuska, dialog prodolzhalsya s momenta perezagruzki.
i copied the code.
run:
Yes, I can see the chat history, but when the window is restarted, the dialog starts over.
and does not continue. For example
i have selected 2 answer buttons from the start and after the window reloads
the dialogue starts with the first question.

Perhaps you can specify an example where this works correctly, so that after restarting, the dialogue continues from the moment of restart.

Please <3
this project is very important to me.

**Comment by dmitrizzle** [comment]
Hey @Iilinkasa, I had another look at your use case. It seems that some of the options you have are:

1) Create an elaborate implementation
2) Update the core `chat-bubble` component (a PR is appreciated!) to improve `options.recallInteractions` method
3) If you aren't willing/able to put effort into either of the above, I suggest you look for another chat UI library.

***

If you're still exploring options 1 or 2, here are some things I found today that may give you insights:


For option 1: I was able to restart the conversation by replacing `answer` keys with function names and then having those functions store some data in `localStorage` before triggering the next conversation block, like this:

```javascript
  var bananaFunction = function () {
        chatWindow.talk(convo, "banana");
        localStorage.setItem("chat-bubble-last-interaction-func", "banana");
      };
```
The conversation could then be restarted on window load like this:
```javascript
chatWindow.talk(
        convo,
        localStorage.getItem("chat-bubble-last-interaction-func") || "ice"
      );
```
However, as mentioned above, the user responses aren't shown in the "recalled" interactions and the mere act of restarting the conversation incorrectly creates an entry in the recalled conversation list. So there's a lot more work that'll need to be done here.

For option 2: There are some issues with the `options.recallInteractions`. I found that it does not seem to remember user responses when they are buttons and it does not make an easy way for restarting the conversation from the previous spot. Overall, the `recallInteractions` feature seems quite buggy and perhaps that's the reason it's not in the docs. Properly fixing this feature should solve your issue and improve the library.

This is all I can do for now.
